ref(processing): Refactor code

armenzg opened the pull request:
This improves the readability and prepares us for further code changes.

stephanie-anderson committed:
ref(getting-started-docs): Add docs for node/serverlesscloud (#53274)

stephanie-anderson committed:
ref(getting-started-docs): Add docs for node/connect (#53270)

closes(https://github.com/getsentry/sentry/issues/52197)

evanpurkhiser committed:
ref(ts): Convert sentry-test/index to tsx (#53314)

dependabot[bot] committed:
chore(deps): bump grpcio from 1.48.1 to 1.56.0 (#52406)

Swatinem committed:
Instrument a couple of assemble-related functions (#53330)

This should make performance transactions for the `assemble_*` tasks a
lot more useful.

Right now, there is only auto-instrumentation for SQL queries, which is
pretty useless without knowing which function they are in.

stephanie-anderson committed:
ref(getting-started-docs): Add docs for node/koa (#53258)

closes https://github.com/getsentry/sentry/issues/52195

nhsiehgit committed:
nit: add stats/loggers for authorize api (#53324)

vuluongj20 committed:
fix(guideAnchor): Fix background in dark mode (#52999)

**Before ——**
<img width="321" alt="image"
src="https://github.com/getsentry/sentry/assets/44172267/75cd6ea2-1a01-4eaa-bf97-a936bc6498a5">


**After ——**
<img width="321" alt="image"
src="https://github.com/getsentry/sentry/assets/44172267/2ed52d46-788d-49fc-a0e4-92cc1cd4c3b3">

mrduncan committed:
fix(hc): Stabilize OrganizationIndex2faTest (#53247)

JonasBa committed:
chore(sentry): bump @sentry packages (#53336)

Co-authored-by: Kev <6111995+k-fish@users.noreply.github.com>

armenzg committed:
ref(processing): Some more refactoring

This helps with the understanding of the code and further changes.

roggenkemper committed:
feat(source-map-debug): Make debug functions independent (#53306)

this pr moves out a lot of the source map debug helper functions into a
new file so that we can use it in a new endpoint we create.
Functionality has been maintained so the old endpoint will output the
expected result.

Closes: https://github.com/getsentry/sentry/issues/53166

scttcper committed:
test(ui): Convert issues list tests to typescript (#53301)

armenzg committed:
get_frames()

armenzg commented on src/sentry/event_manager.py:
```diff
@@ -462,9 +462,6 @@ def save_error_events(
         is_reprocessed = is_reprocessed_event(job["data"])
 
-        # This metric can be used to track how many error events there are per platform
-        metrics.incr("save_event.error", tags={"platform": job["event"].platform or "unknown"})
```
We now have this metric instead:
https://github.com/getsentry/sentry/blob/c2854dc5ed7ef766b71b263aa70a7b63fdb28933/src/sentry/event_manager.py#L444

lobsterkatie replied:
If they do the same thing, the new one is indeed better, because `save_event.error` makes it sound like there was a problem saving the event, not that what we're doing is successfully saving an error event.

armenzg commented on src/sentry/stacktraces/processing.py:
```diff
@@ -33,6 +33,9 @@ def __eq__(self, other: object) -> bool:
         return self is not other
 
+    def get_frames(self) -> Sequence[Dict[str, Any]] | None:
+        return get_path(self.stacktrace, "frames", filter=True, default=())
```
This call was happening all throughout this file.

lobsterkatie replied:
Perhaps this is a moot point given other refactors you have coming up, but if you made this a helper function (rather than a `StacktraceInfo` method), you could use it in places like `_append_stacktrace` also.

azaslavsky committed:
ref(backup): Do deepcopy before validation (#53346)

Previously, all validation mutated the inputs in-place. This was
performant (no need to make deep copies of two potentially large JSON
blobs), but potentially bug prone if not held carefully, since the
validation itself mutated the inputs via JSONScrubbingComparators. This
means that the post-comparison input objects MUST NOT be re-used, since
they are meaningfully different from the inputs that were declared
valid.

Instead, we now do deepcopies under the hood, leaving the originally
passed-in inputs untouched.

Issue: getsentry/team-ospo#156

armenzg committed:
Minor fixes

armenzg commented on src/sentry/stacktraces/processing.py:
```diff
@@ -192,16 +195,15 @@ def find_stacktraces_in_data(
-            frame.get("platform") or data.get("platform")
-            for frame in get_path(stacktrace, "frames", filter=True, default=())
-        }
+        platforms = _get_frames_metadata(frames, data.get("platform", "unknown"))
```
This is in preparation of my next PR.

lobsterkatie replied:
So this function will eventually get more than just the platforms?

armenzg replied:
Correct.

armenzg commented on src/sentry/stacktraces/processing.py:
```diff
@@ -192,16 +195,15 @@ def find_stacktraces_in_data(
             StacktraceInfo(
                 stacktrace=stacktrace,
+                # XXX: This contains all the contents of stacktrace
                 container=container,
```
In the future, we should extract the frames directly from the container rather than from stacktrace or even determine why we need a container at all.

lobsterkatie replied:
There are certain platforms where we need the container even if there's nothing in it (or nothing that's not `None` or empty). See https://github.com/getsentry/sentry/pull/15795#discussion_r349916876.

armenzg commented on src/sentry/stacktraces/processing.py:
```diff
@@ -257,17 +268,14 @@ def normalize_stacktraces_for_grouping(data: Any, grouping_config: Any = None) -
 
     stacktrace_frames = []
-    stacktrace_exceptions = []
+    stacktrace_container = []
```
A more consistent naming.

Eventually I would like to iterate over a list of StacktraceInfo. No need for two data structures.

lobsterkatie replied:
```suggestion
    stacktrace_containers = []
```

asottile-sentry committed:
ref: future-proof using signature modification (#53347)

the signature of a few of these changes in newer `django-stubs` -- this
makes the plugin a little more future-proof

armenzg committed:
Last one

asottile-sentry committed:
ref: remove unused pickle complain setting (#53349)

<!-- Describe your PR here. -->

gggritso committed:
chore(ts): Convert a few more test stubs (#53343)

- Remove unused stub
- TestStubs.EventsStats
- TestStubs.DetailedEvents
- TestStubs.Event
- TestStubs.EventIdQueryResult
- TestStubs.EventEntryStacktrace
- TestStubs.EventEntryDebugMeta
- TestStubs.EventAttachment
- TestStubs.DiscoverSavedQuery
- TestStubs.DocIntegration
- TestStubs.EventEntry
- TestStubs.DebugFile
- TestStubs.Commit
- TestStubs.CodeOwner

cathteng committed:
fix(github-comments): check error.json exists (#53344)

Fixes https://sentry.sentry.io/issues/4330699256/?project=1

dashed committed:
feat(hybrid-cloud): Add integration proxy to AWS Lambda client (#53038)

scttcper committed:
feat(issues): Display semver info when resolved in release (#53243)

john-z-yang committed:
fix(test): Add missing __init__.py (#53352)

Add missing `__init__.py`

cathteng committed:
ref(github-growth): check external id when creating repo (#53307)

armenzg committed:
Minor change

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -257,17 +268,14 @@ def normalize_stacktraces_for_grouping(data: Any, grouping_config: Any = None) -
-            frames = get_path(stacktrace_info.stacktrace, "frames", filter=True, default=())
-            if frames:
-                stacktrace_frames.append(frames)
-                stacktrace_exceptions.append(
-                    stacktrace_info.container if stacktrace_info.is_exception else None
-                )
```
I think this has to do with https://github.com/getsentry/sentry/pull/15795#discussion_r349916876 as well, so I'd be wary of changing it unless you've investigated how it would affect the problems that PR was trying to solve.

asottile-sentry committed:
ref: use named arguments for drf serializers (#53354)

see https://github.com/encode/django-rest-framework/pull/7632




<!-- Describe your PR here. -->

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -220,13 +222,22 @@ def _append_stacktrace(stacktrace: Any, container: Any, is_exception: bool = Fal
 
     if include_raw:
-        for info in rv[:]:
-            if info.container is not None:
```
This is another spot where https://github.com/getsentry/sentry/pull/15795#discussion_r349916876 might come into play. IDK if that use case requires empty dictionaries to be included.

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -192,16 +195,15 @@ def find_stacktraces_in_data(
         rv.append(
             StacktraceInfo(
                 stacktrace=stacktrace,
+                # XXX: This contains all the contents of stacktrace
```
I'm not sure what this comment means. The container is going to be the full dictionary with `type`, `value`, `stacktrace`, `raw_stacktrace`, and `mechanism` (and possible more).

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -33,6 +33,9 @@ def __eq__(self, other: object) -> bool:
     def __ne__(self, other: object) -> bool:
         return self is not other
 
+    def get_frames(self) -> Sequence[dict[str, str]]:
```
```suggestion
    def get_frames(self) -> Sequence[dict[str, str | int | bool]]:
```

Unless we stringify everything at some point - do we?

(I'm basing this on frames like this one:

```json
{
    "function": "c",
    "filename": "app:///_next/static/chunks/main-aadfe6c52ca3d609.js",
    "abs_path": "app:///_next/static/chunks/main-aadfe6c52ca3d609.js",
    "lineno": 1,
    "colno": 1379,
    "in_app": false
},
```
)

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -33,6 +33,9 @@ def __eq__(self, other: object) -> bool:
         return self is not other
 
+    def get_frames(self) -> Sequence[dict[str, str]]:
+        return get_path(self.stacktrace, "frames", filter=True, default=())
```
Total nit: If normally this gives back a list, is there a reason to make the default a tuple rather than a list?

schew2381 committed:
ref(api): Create basic auth header helper function (#53296)

Formatting a basic auth header is long and annoying to type out so create util function to handle instead

lobsterkatie commented on src/sentry/stacktraces/processing.py:
```diff
@@ -257,17 +268,15 @@ def normalize_stacktraces_for_grouping(data: Any, grouping_config: Any = None) -
+    stacktrace_container = []
 
     with sentry_sdk.start_span(op=op, description="find_stacktraces_in_data"):
+        # XXX: Why do we want include the raw stacktrace?
```
I have been presuming it's so that we can run `_normalize_in_app` on them as well.

michellewzhang committed:
feat(replays): add initial replay cards to replay list (#53323)

closes #53086

created two experimental tables on the replay list page: 
1. top 3 most erroneous replays within the last 7 days
2. top 3 most dead click replays within the last 7 days
(can change these time periods, I just thought 1 day was too little time
for a useful sample size since a lot of numbers showed 0)

notes:
- these tables are static when compared to the big replay list table, so
pagination doesn't affect them (see below for screenshots showing this)
- in theory, rage clicks always < dead clicks, which is why I chose to
sort by dead clicks

page 1:
<img width="1250" alt="SCR-20230721-jybf"
src="https://github.com/getsentry/sentry/assets/56095982/1c2f6981-e504-4d90-9f88-4c4dfc4523f5">


page 2 (same tables, different replays on big list):
<img width="1238" alt="SCR-20230721-jxge"
src="https://github.com/getsentry/sentry/assets/56095982/d4b231e3-8786-4144-8122-cc57af0ef3ca">

asottile-sentry committed:
ref: hack around mypy's mistreatment of descriptors in unions (#53308)

working around https://github.com/python/mypy/issues/5570

this is especially important as more things are being moved to unions of
`HC | Model`

asottile-sentry committed:
ref: fix import to pytz (#53359)

django 4.x does not have this attribute at all -- in 3.x this is an
import of `pytz`

scttcper committed:
fix(issues): Correctly use `issue-release-semver` flag (#53356)

DominikB2014 committed:
fix(starfish): remove operation from span summary (#53358)

DominikB2014 committed:
fix(starfish): remove table column in db module (#53357)

shruthilayaj committed:
feat(starfish): Plot multiple releases on mobile service view (#53335)

![Screenshot 2023-07-21 at 10 50 47
AM](https://github.com/getsentry/sentry/assets/63818634/d97dde14-2f3c-41eb-9f3e-17e753c0f101)

asottile-sentry committed:
ref: don't guess about str / bytes in tests (#53366)

force_str / force_bytes really should not be used -- these are the easy
ones to fix




<!-- Describe your PR here. -->

markstory committed:
feat: Refactor and add tests for load-mock (#53302)

Refactor load-mocks into a set of smaller functions, and have some high
level tests to ensure that load-mocks will run. These tests aren't
exhaustive but they should help us catch and prevent regressions in
load-mocks in the future.

Zylphrex committed:
chore(profiling): Remove profiling-ga flag from frontend (#53360)

Profiling has been in GA for a while now, time to remove the badges.

Commit:
ref(crons): Center placeholder timeline view (#53376)

Commit:
feat(stacktrace): Add toggle on Native stacktrace for non-'In App' frames (#53313)

Closes https://github.com/getsentry/sentry/issues/53300

Add toggle for show/hide x more frame(s) on stacktrace for hidden
non-inApp frames.
![Screenshot 2023-07-20 at 3 42 18
PM](https://github.com/getsentry/sentry/assets/22582037/840b7202-1436-4709-80d7-ce1036915651)
![Screenshot 2023-07-20 at 3 42 28
PM](https://github.com/getsentry/sentry/assets/22582037/aec77325-9639-4ac9-8344-d00e14c338ae)

asottile-sentry committed:
ref: use an explicit primary key for social_auth model (#53350)

this was the implicit default value before -- django 3.2 makes it warn,
django 4 makes it an error

social_auth seems to be the only one receiving this default value as we
set `id = BoundedBigAutoField(primary_key=True)` in our base model

azaslavsky committed:
test(backup): Ensure coverage of all exportable models (#53295)

The basic idea here is to include a test that finds all non-abstract
descendants of our `BaseModel`, filter down to just the exportable ones
(ie, those that set `__include_in_export__ = True`), and then ensure
that each of them is included in at least one of the tests seen in
`test_models.py`. This is done by introducing a new `mark` wrapper class
which ingests all of the "target" models used by the `@targets` test
decorator at init time, allowing us to create an exhaustive list of all
types passed to `@targets`.

Issue: getsentry/team-ospo#156

scttcper committed:
fix(Issues): Switch from "timestamp" to "non-semver" (#53383)

schew2381 committed:
fix(ui): Align pending request and request access button (#53386)

Before:
<img width="1264" alt="Screenshot 2023-07-21 at 3 24 40 PM"
src="https://github.com/getsentry/sentry/assets/67301797/8a664fb5-3faa-44a1-9f09-2cc940fac461">

After:
<img width="1248" alt="Screenshot 2023-07-21 at 3 25 12 PM"
src="https://github.com/getsentry/sentry/assets/67301797/6a3cf1e8-4ad7-4534-91c3-9a1b28baf53d">

azaslavsky committed:
test(backup): Add more thorough model coverage checks (#53298)

We've already introduce model coverage checks in a previous PR, but
these only validated that every `__include_in_export = True`-marked
model is exercised by at least one test in
`tests/sentry/backup/test_models.py`. This change improves on that
slightly, by only considering a model to be "tested" if the matching
instance in the output has all of its fields set, and none of those
fields are set to their default value (if they have one). This ensures
that not only are we covering every exportable model, we are covering
every field of every model.

Issue: getsentry/team-ospo#156

barkbarkimashark committed:
ref(alert): swap the internal vs comparisonInterval when displaying alert rules … (#53387)

The FE alert rule display text for percent comparisons seems to have the
interval and comparisonInterval reversed.

Resolves https://github.com/getsentry/sentry/issues/53382

dashed committed:
chore(hybrid-cloud): Mark Discord integration test as stable (#53381)

dashed committed:
chore(hybrid-cloud): Mark Bitbucket webhook endpoint as region silo (#53380)

corps committed:
ref(hc): Make the auth middleware use hybrid cloud authentication. (#53285)

Uses hybrid cloud level authentication in django views by default.

corps committed:
fix(hc): Support split db monolith mode via smarter delegators (#53365)

Monolith mode delegators now *also* pick their implementation by
attempting to examine any open transactions and selecting for an
implementation that would align with the currently open tranaction.

This fixes the subtle issue for hybrid cloud services like the
log_service, as well as some helper methods on the organization_service,
that do not invoke RPC, but rather, *pick different models* locally
based on the silo mode. For Monolith mode, this means picking one model
that works on one db, even if being invoked inside the transaction of
the other db, threatening the transactional atomicity that is desired.
(`schedule_signal`, for instance, hopes to leave an outbox behind in the
same transaction as invocation).

The Sentry Options backend *may* also need this, but unclear. Going to
analyze split db tests after this lands and follow up.

getsentry-bot committed:
Revert "fix(hc): Support split db monolith mode via smarter delegators (#53365)"

This reverts commit 7b0a6922e5e1f84258f3f0b2664fcc85cbc5b6be.

Co-authored-by: corps <593850+corps@users.noreply.github.com>

mrduncan committed:
fix(hc): Add stable=True to more tests (#52656)

Co-authored-by: Zach Collins <recursive.cookie.jar@gmail.com>
Co-authored-by: Zachary Collins <zachary.collins@sentry.io>

corps committed:
ref(hc): Fold PagerDutyService objects into OrganizationIntegration config directly. (#53342)

1. Switch back PagerDutyService to control silo now that getsentry is
updated
2. On save, put pagerdutyservice configuration into the organization
integration config object
3. Change sentry read paths to use this
4. Migrate all PagerDutyService objects to backfill the org integrations

corps committed:
fix(hc): Support split db monolith mode via smarter delegators (#53395)

Same as https://github.com/getsentry/sentry/pull/53365/files but
conditionalizing the testutil import to prevent issue in prod.

priscilawebdev committed:
fix(getting-started-docs): Fix duplicated translation (#53408)

romtsn committed:
ref(getting-started-docs): Migrate android wizard to the main sentry repo (#53351)

Commit:
feat(sourcemaps): Add more indexes to tables for artifact bundles (#51876)

Swatinem committed:
Validate assemble_artifact arguments earlier (#53333)

- Ensures that we always have a `Project` for ArtifactBundles
- Ensures that there is always a release for a Releasebundle
- Defaults the `bundle_id` earlier

Co-authored-by: Riccardo Busetti <riccardo.busetti@sentry.io>

obostjancic committed:
fix(alerts): skip on demand for standard alerts (#53416)

obostjancic committed:
fix(alerts): match interval to granularity (#53415)

untitaker committed:
ref: Remove dead code (#52876)

Commit:
ref(dynamic-sampling): Simplify boost low volume transactions calculation (#53418)

armenzg committed:
Update src/sentry/stacktraces/processing.py

Co-authored-by: Katie Byers <katie.byers@sentry.io>

armenzg committed:
Address feedback
